File: String/longest-palindromic-substring/longest_palindromic_substring.py
#  Project : SDE Placement Questions
#  Filename : longest_palindromic_substring.py
#  Author : thameem
#  Current modification time : Sat, 25 Mar 2023 at 2:41 pm India Standard Time
#  Last modified time : Sat, 25 Mar 2023 at 2:41 pm India Standard Time


def check_palindrome(l, r, len_of_string, s, res_length, res):
    while l >= 0 and r < len_of_string and s[l] == s[r]:
        if (r - l + 1) > res_length:
            res_length = r - l + 1
            res = s[l:r + 1]

        l -= 1
        r += 1

    return res, res_length


def longest_palindrome(s: str) -> str:
    res = ""

    res_length = 0

    len_of_string = len(s)

    for i in range(len_of_string):
        res, res_length = check_palindrome(i, i, len_of_string, s, res_length, res)
        res, res_length = check_palindrome(i, i + 1, len_of_string, s, res_length, res)

    return res


if __name__ == '__main__':
    print(longest_palindrome(s="babad"))

File: String/longest-palindromic-substring/longest_palindromic_substring.cpp
#include <iostream>

using namespace std;

void check_palindrome(int l,int r,string &res,string s, int &res_len, int &s_len)
    {
        while(l>=0 && r<s_len && s[l]==s[r])
        {
            if((r-l+1)>res_len)
            {
                res_len = r-l+1;
                res = s.substr(l,res_len);
            }

            l--;
            r++;
        }

    }

string longestPalindrome(string s)
{
    string res = "";
    int res_len = 0;
    int s_len = s.length();

    for(int i=0;i<s_len;i++)
    {

        check_palindrome(i,i,res,s,res_len,s_len);
        check_palindrome(i,i+1,res,s,res_len,s_len);

    }

    return res;
}

int main()
{
    cout<<longestPalindrome("b")<<"\n";
    return 0;
}